invokeai/app/services/config.py

```diff
@@ -0,0 +1,364 @@
+        
+        # Set the runtime root directory. We parse command-line switches here
+        # in order to pick up the --root_dir option.
+        self.parse_args(argv)
+        if not conf:
+            try:
+                conf = OmegaConf.load(self.root_dir / INIT_FILE)
+            except:
+                pass
+        InvokeAISettings.initconf = conf
+
+        # parse args again in order to pick up settings in configuration file
+        self.parse_args(argv)
+
+        # restore initialization values
+        hints = get_type_hints(self)
+        for k in kwargs:
+            setattr(self,k,parse_obj_as(hints[k],kwargs[k]))
+
+    @property
+    def root_path(self)->Path:
+        '''
+        Path to the runtime root directory
+        '''
+        if self.root:
+            return self.root.expanduser()
+        else:
+            return self.find_root()
+
+    @property
+    def root_dir(self)->Path:
+        '''
+        Alias for above.
+        '''
+        return self.root_path
+
+    def _resolve(self,partial_path:Path)->Path:
+        return (self.root_path / partial_path).resolve()
+
+    @property
+    def output_path(self)->Path:
+        '''
+        Path to defaults outputs directory.
+        '''
+        return self._resolve(self.outdir)
+
+    @property
+    def model_conf_path(self)->Path:
+        '''
+        Path to models configuration file.
+        '''
+        return self._resolve(self.conf_path)
+
+    @property
+    def conf(self)->Path:
+        '''
+        Path to models configuration file (alias for model_conf_path).
+        '''
+        return self.model_conf_path
+
+    @property
+    def embedding_path(self)->Path:
+        '''
+        Path to the textual inversion embeddings directory.
+        '''
+        return self._resolve(self.embedding_dir) if self.embedding_dir else None
+    
+    @property
+    def lora_path(self)->Path:
+        '''
+        Path to the LoRA models directory.
+        '''
+        return self._resolve(self.lora_dir) if self.lora_dir else None
+
+    @property
+    def autoconvert_path(self)->Path:
+        '''
+        Path to the directory containing models to be imported automatically at startup.
+        '''
+        return self._resolve(self.autoconvert_dir) if self.autoconvert_dir else None
+
+    @property
+    def gfpgan_model_path(self)->Path:
+        '''
+        Path to the GFPGAN model.
+        '''
+        return self._resolve(self.gfpgan_model_dir) if self.gfpgan_model_dir else None
+
+    @staticmethod
+    def find_root()->Path:
+        '''
+        Choose the runtime root directory when not specified on command line or
+        init file.
+        '''
+        return _find_root()
+
+class InvokeAIWebConfig(InvokeAIAppConfig):
+    '''
+    Web-specific settings
+    '''
+    #fmt: off
+    type               : Literal["web"] = "web"
+    allow_origins      : List = Field(default=[], description="Allowed CORS origins")
```

Comment:
These should maybe all be grouped under a CORS category, or with `cors` in the name?

Reply:
> Trying out web UI: invokeai.py --web ==> works well (though only tested TextToImage) invokeai-new.py --web ==> server crashed on startup with error:
> 
> ` File "...../invokeai/app/services/graph.py", line 1179, in validate_exposed_nodes graph = values['graph'] KeyError: 'graph' ERROR: Application startup failed. Exiting.`
> 
> It's possible there's something weird in my setup that's triggering this only for me, but I don't know what that would be.
> 
> I also ran pytest for the three test files modified in this PR, and they passed all tests.
> 
> On Ubuntu 22.04, Nvidia A5000 card

Is this behavior specific to this PR, or is it present in the base `main` branch as well? The web server is fairly unstable at the moment.

Reply:
Done:
``` 
invokeai-node-web --help
[... all the config stuff in the previous comment...]

Features:
  --nsfw_checker, --no-nsfw_checker
                        Enable/disable the NSFW checker (default: True)
  --restore, --no-restore
                        Enable/disable face restoration code (default: True)
  --esrgan, --no-esrgan
                        Enable/disable upscaling code (default: True)
  --patchmatch, --no-patchmatch
                        Enable/disable patchmatch inpaint code (default: True)
  --internet_available, --no-internet_available
                        If true, attempt to download models on the fly; otherwise only use local models (default: True)
  --log_tokenization, --no-log_tokenization
                        Enable logging of parsed prompt tokens. (default: False)

Cross-Origin Resource Sharing:
  --allow_origins ALLOW_ORIGINS
                        Allowed CORS origins
  --allow_credentials, --no-allow_credentials
			Allow CORS credentials (default: True)
  --allow_methods ALLOW_METHODS
                        Methods allowed for CORS
  --allow_headers ALLOW_HEADERS
                        Headers allowed for CORS

Web Server:
  --host HOST           IP address to bind to
  --port PORT           Port to bind to
```